### PR TITLE
Support complex numbers in `cupy.linalg.qr()`

### LINKED ISSUE #2193: cupy.linalg.qr() of complex matrix gives wrong answer
cupy.linalg.qr() works well for real matrix but returns wrong result for complex ones.  

```py
>>> import cupy as cp
>>> x = cp.random.randn(3,2)+cp.random.randn(3,2)*1.0j
>>> x
array([[-0.24066196+1.51554653j,  0.78608315-0.07083773j],
       [-0.22975259-0.27401681j, -0.94060383-1.28354323j],
       [ 0.28654721-0.34910445j, -2.5263385 -0.25618918j]])
>>> q,r=cp.linalg.qr(x)
>>> q
array([[-0.15510171+0.9767387j ,  0.78608315-0.07083773j],
       [-0.14807084-0.57906488j, -0.94060383-1.28354323j],
       [-0.21132318-0.78741754j, -2.5263385 -0.25618918j]])
>>> r
array([[ 1.55163969-0.84558675j,  0.78608315-0.07083773j],
       [ 0.        +0.j        , -0.94060383-1.28354323j]])
>>> xp = cp.einsum('ij,jk->ik',q,r)
>>> xp
array([[ 0.58525534+1.64669848j, -0.883049  -0.16355659j],
       [-0.71940218-0.77329331j, -0.9201633 +1.96990723j],
       [-0.99372727-1.04309622j,  1.82555733+2.87963121j]])
>>> cp.linalg.norm(xp-x)
array(6.7425088)
```

In addition, both q and r are different from that of numpy.linalg.qr() for the same matrix.

* Conditions
  - CuPy version: 5.4.0
  - OS/Platform: linux
  - CUDA version: 10.0.130
  - cuDNN/NCCL version (if applicable): 7.3.1

### PR DESCRIPTION
**Pull request by leofang**
Close #2193. cf. #2468 & #2509.

### CODE REVIEW

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Comment by leofang** [comment]
@kmaehashi the Jenkins error is due to a recent update in `attrs`, see https://github.com/pytest-dev/pytest/issues/3280#issuecomment-538031871.

**Comment by toslunar** [comment]
@leofang Thanks for reporting and investigating the issue.  It's fixed by chainer/chainer-test#529.

**Comment by leofang** [comment]
@toslunar No problem. I encountered the very same issue during a local test just a day earlier, so I knew what exactly the cause was.

@kmaehashi any chance you could kick off Jenkins again? Thanks.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by leofang** [comment]
Ready to go?

File: cupy/linalg/decomposition.py [hunk 1]
```diff
@@ -118,21 +118,29 @@ def qr(a, mode='reduced'):
...
+    elif dtype == 'F':
+        geqrf_bufferSize = cusolver.cgeqrf_bufferSize
+        geqrf = cusolver.cgeqrf
+    elif dtype == 'D':
+        geqrf_bufferSize = cusolver.zgeqrf_bufferSize
+        geqrf = cusolver.zgeqrf
+    else:
+        msg = ('dtype must be float32, float64, complex64 or complex128'
+               ' (actual: {})'.format(a.dtype))
+        raise ValueError(msg)
```
Comment (toslunar) [inline cupy/linalg/decomposition.py]: Nitpick: The block is unreachable.  (See https://github.com/cupy/cupy/pull/2509#discussion_r331863033)

**Comment by leofang** [comment]
Thanks for help, @toslunar and @kmaehashi!

**Pull request merged**